types/coin.go

```diff
@@ -508,15 +513,49 @@ func ParseCoin(coinStr string) (coin Coin, err error) {
 
 	denomStr, amountStr := matches[2], matches[1]
 
-	amount, ok := NewIntFromString(amountStr)
-	if !ok {
-		return Coin{}, fmt.Errorf("failed to parse coin amount: %s", amountStr)
-	}
-
 	if denomStr != strings.ToLower(denomStr) {
 		return Coin{}, fmt.Errorf("denom cannot contain upper case characters: %s", denomStr)
 	}
 
+	// test if the amount is specified using scientific notation
+	scinMatches := reScinAmt.FindStringSubmatch(amountStr)
+
+	ok := false
+	var amount Int;
+
+	if len(scinMatches) == 4 {
+
+		// Parse scientific notation such as 1.23e+12stake
+		var f float64
+		var dec int
+		
+		if f, err = strconv.ParseFloat(scinMatches[1], 64); err != nil {
+			return
+		}
+		
+		if dec, err = strconv.Atoi(scinMatches[3]); err != nil {
+			return 
+		}
+
+		// adjust dec by the length of the decimal part in the number
+		d := reDecPart.FindStringSubmatch(scinMatches[1])
+		if len(d) > 1 {
+			l := len(d[1])
+			if l > dec {
+				return Coin{}, fmt.Errorf("amount %s has too many digits after the decimal point resulting in precision loss",amountStr)
+			}			
+			f, dec = f * math.Pow10(l), dec - l
+		}
+		amount, ok = NewIntWithDecimal(int64(f), dec), true
+
+	} else {
+		amount, ok = NewIntFromString(amountStr)
+	}
```

Comment:
to reduce indentation and increase clarity - rather than having a large if-else statement, we should first check if this is _not_ a scientific notation, and perform the old calculations first, and return in a small if statement near line 520... then all the code for dealing with scientific notation can come afterwards and not be indented in an if statement

Reply:
broke down if-else statements to be more granular and bail on errors first